### ISSUE TITLE
[PROMOTION]make model much more generic

### ISSUE THREAD

**Issue by novioleo** [novioleo/crnn.mxnet#3]
Recently I've changed the structure of the web to make it more generic, adapting to more than just generating data sets. The current model can very good fitting we generated data, but the reality data and background too miscellaneous and too much, we can not collect enough data, therefore, in order to improve the ability of generic OCR,I put forward several solutions:
**1. Background filtering:** the operation of max-pooling minus avg-pooling is performed on the first or second layer.
**2. Reduce the number of pooling:** not compress the number of rows to 1 row, and increase the input steps to LSTM.
**3. Reduce the number of kernels of convolution:** it can improve generalization, but it will slow down the convergence rate and reduce the accuracy slightly. BUT It also reduces the size of the model.

**Comment by novioleo** [comment]
@goodtogood Do you have any good ideas?

**Comment by novioleo** [comment]
contain above enhancement in 78eae56

**Comment by xiaomaxiao** [comment]
I  test CRNN in english character , i found  word training is better than randomly gen characters
i also use the method of randomly generating chinese characters  , i found it's too hard for training.
whether you have completed the chinese character training with randomly gen characters?
just 10000 images?

if you apply above 3 solutions ,it's same to tesseract -ocr for  lstm 4.0.

**Comment by novioleo** [comment]
随机中文文字效果很差。而且泛化性很差。但是通过调整filter的个数和整个卷积的结构，能够优化整个识别的效果。我用了80W张中文的随机字符生成的。tesseract-ocr主要是太大了，crnn能够根据我的需求调整模型大小。

**Comment by fendaq** [comment]
@novioleo @xiaomaxiao 两位好，我在搞印刷文本的crnn中文识别 也使用了随机生成中文方式
 生成了大约1000w样本 但是真实数据上测试效果不好 泛化性差 各位有什么好的建议?

**Comment by novioleo** [comment]
@fendaq 泛化性差主要还是样本造成的，CRNN很容易对样本拟合。另外我写的那个sceneText生成只是一个简易的版本。可能还需要根据你的实际情况来调整下。
另外，如果是识别中文的话，卷积层的卷积核数多点好些。

**Comment by xiaomaxiao** [comment]
@fendaq 不要使用随机生成。

**Comment by fendaq** [comment]
@xiaomaxiao 我用opencv 做行检测和切分。我建了一个qq群，加一下，有问题可以一起讨论一下

**Comment by fendaq** [comment]
@novioleo 能否共享一下你优化整个识别的效果的filter和cnn层的调整参数，谢谢

**Comment by novioleo** [comment]
@fendaq 我没做中文的识别。目前做的银行卡的，而且是面向移动端的，这个参数对你没有任何意义。

**Comment by xiaomaxiao** [comment]
@novioleo  我用了100W 中文随机，模型无法收敛~~~

**Comment by novioleo** [comment]
@xiaomaxiao 多少epoch

**Comment by xiaomaxiao** [comment]
@novioleo  2 epoch ， loss 都在 50 不降了

**Comment by novioleo** [comment]
@xiaomaxiao 你的训练参数po上来看下

**Comment by xiaomaxiao** [comment]
`       
input = Input(shape=(img_h,None,1),name='the_input')
m = Conv2D(64,kernel_size=(3,3),activation='relu',padding='same',name='conv1')(input)
m = MaxPooling2D(pool_size=(2,2),strides=(2,2),name='pool1')(m)
m = Conv2D(128,kernel_size=(3,3),activation='relu',padding='same',name='conv2')(m)
m = MaxPooling2D(pool_size=(2,2),strides=(2,2),name='pool2')(m)
m = Conv2D(256,kernel_size=(3,3),activation='relu',padding='same',name='conv3')(m)
m = BatchNormalization(axis=1)(m)
m = Conv2D(256,kernel_size=(3,3),activation='relu',padding='same',name='conv4')(m)

m = ZeroPadding2D(padding=(0,1))(m)
m = MaxPooling2D(pool_size=(2,2),strides=(2,1),padding='valid',name='pool3')(m)

m = Conv2D(512,kernel_size=(3,3),activation='relu',padding='same',name='conv5')(m)
m = BatchNormalization(axis=1)(m)
m = Conv2D(512,kernel_size=(3,3),activation='relu',padding='same',name='conv6')(m)

m = ZeroPadding2D(padding=(0,1))(m)
m = MaxPooling2D(pool_size=(2,2),strides=(2,1),padding='valid',name='pool4')(m)
m = Conv2D(512,kernel_size=(2,2),activation='relu',padding='valid',name='conv7')(m)
m = BatchNormalization(axis=1)(m)

m = Permute((2,1,3),name='permute')(m)
m = TimeDistributed(Flatten(),name='timedistrib')(m)

m = Bidirectional(GRU(rnnunit,return_sequences=True,implementation=2),name='blstm1')(m)
#m = Bidirectional(LSTM(rnnunit,return_sequences=True),name='blstm1')(m)
m = Dense(rnnunit,name='blstm1_out',activation='linear')(m)
#m = Bidirectional(LSTM(rnnunit,return_sequences=True),name='blstm2')(m)
m = Bidirectional(GRU(rnnunit,return_sequences=True,implementation=2),name='blstm2')(m)
y_pred = Dense(nclass,name='blstm2_out',activation='softmax')(m)

basemodel = Model(inputs=input,outputs=y_pred)
basemodel.summary()


labels = Input(name='the_labels',shape=[maxlabellength],dtype='float32')
input_length = Input(name='input_length', shape=[1], dtype='int64')
label_length = Input(name='label_length', shape=[1], dtype='int64')

loss_out = Lambda(ctc_lambda_func, output_shape=(1,), name='ctc')([y_pred, labels, input_length, label_length]) 

model = Model(inputs=[input, labels, input_length, label_length], outputs=loss_out)


adam = Adam(lr=0.001,beta_1=0.9,beta_2=0.999)
#adadelta = Adadelta()
model.compile(loss={'ctc': lambda y_true, y_pred: y_pred}, optimizer=adam,metrics=['accuracy'])

`
`Layer (type)                 Output Shape              Param #   
=================================================================
the_input (InputLayer)       (None, 32, None, 1)       0         
_________________________________________________________________
conv1 (Conv2D)               (None, 32, None, 64)      640       
_________________________________________________________________
pool1 (MaxPooling2D)         (None, 16, None, 64)      0         
_________________________________________________________________
conv2 (Conv2D)               (None, 16, None, 128)     73856     
_________________________________________________________________
pool2 (MaxPooling2D)         (None, 8, None, 128)      0         
_________________________________________________________________
conv3 (Conv2D)               (None, 8, None, 256)      295168    
_________________________________________________________________
batch_normalization_1 (Batch (None, 8, None, 256)      32        
_________________________________________________________________
conv4 (Conv2D)               (None, 8, None, 256)      590080    
_________________________________________________________________
zero_padding2d_1 (ZeroPaddin (None, 8, None, 256)      0         
_________________________________________________________________
pool3 (MaxPooling2D)         (None, 4, None, 256)      0         
_________________________________________________________________
conv5 (Conv2D)               (None, 4, None, 512)      1180160   
_________________________________________________________________
batch_normalization_2 (Batch (None, 4, None, 512)      16        
_________________________________________________________________
conv6 (Conv2D)               (None, 4, None, 512)      2359808   
_________________________________________________________________
zero_padding2d_2 (ZeroPaddin (None, 4, None, 512)      0         
_________________________________________________________________
pool4 (MaxPooling2D)         (None, 2, None, 512)      0         
_________________________________________________________________
conv7 (Conv2D)               (None, 1, None, 512)      1049088   
_________________________________________________________________
batch_normalization_3 (Batch (None, 1, None, 512)      4         
_________________________________________________________________
permute (Permute)            (None, None, 1, 512)      0         
_________________________________________________________________
timedistrib (TimeDistributed (None, None, 512)         0         
_________________________________________________________________
blstm1 (Bidirectional)       (None, None, 512)         1181184   
_________________________________________________________________
blstm1_out (Dense)           (None, None, 256)         131328    
_________________________________________________________________
blstm2 (Bidirectional)       (None, None, 512)         787968    
_________________________________________________________________
blstm2_out (Dense)           (None, None, 3859)        1979667   
=================================================================
Total params: 9,628,999
Trainable params: 9,628,973
Non-trainable params: 26`

batchsize 64 , adam lr=0.001

**Comment by novioleo** [comment]
filter数全部弄成128，batch size调大。另外你的字符数是多少。你的最大宽度最好是最多字符数4倍

**Comment by xiaomaxiao** [comment]
@novioleo  3800字符 一个图片 7个字 ，可能是宽度太大 难以收敛

**Comment by novioleo** [comment]
你先改吧。
